server/sonar-server/src/main/java/org/sonar/server/test/ws/TestsListAction.java

```diff
@@ -0,0 +1,221 @@
+
+    JsonWriter json = response.newJsonWriter().beginObject();
+    writeTests(tests.getDocs(), componentsByTestFileUuid, json);
+    searchOptions.writeJson(json, tests.getTotal());
+    json.endObject().close();
+  }
+
+  private void writeTests(List<TestDoc> tests, Map<String, ComponentDto> componentsByTestFileUuid, JsonWriter json) {
+    json.name("tests").beginArray();
+    for (TestDoc test : tests) {
+      String fileUuid = test.fileUuid();
+      json.beginObject();
+      json.prop("testUuid", test.testUuid());
+      json.prop("fileUuid", fileUuid);
+      json.prop("name", test.name());
+      json.prop("status", test.status());
+      json.prop("durationInMs", test.durationInMs());
+      json.prop("message", test.message());
+      json.prop("stacktrace", test.stackTrace());
+      json.prop("coveredLines", coveredLines(test.coveredFiles()));
+      json.prop("fileKey", componentsByTestFileUuid.get(fileUuid).key());
+      json.prop("fileLongName", componentsByTestFileUuid.get(fileUuid).longName());
+      json.endObject();
+    }
+    json.endArray();
+  }
+
+  private long coveredLines(List<CoveredFileDoc> coveredFiles) {
+    long numberOfLinesCovered = 0L;
+    for (CoveredFileDoc coveredFile : coveredFiles) {
+      numberOfLinesCovered += coveredFile.coveredLines().size();
+    }
+
+    return numberOfLinesCovered;
+  }
+
+  private Map<String, ComponentDto> buildComponentsByTestFileUuid(List<TestDoc> tests) {
+    List<String> fileUuids = Lists.transform(tests, new NonNullInputFunction<TestDoc, String>() {
+      @Override
+      protected String doApply(TestDoc testDoc) {
+        return testDoc.fileUuid();
+      }
+    });
+    DbSession dbSession = dbClient.openSession(false);
+    List<ComponentDto> components;
+    try {
+      components = dbClient.componentDao().getByUuids(dbSession, fileUuids);
+    } finally {
+      MyBatis.closeQuietly(dbSession);
+    }
+
+    return Maps.uniqueIndex(components, new Function<ComponentDto, String>() {
+      @Override
+      public String apply(@Nullable ComponentDto componentDto) {
+        return componentDto.uuid();
+      }
+    });
+  }
+
+  private SearchResult<TestDoc> searchTests(@Nullable String testUuid, @Nullable String testFileUuid, @Nullable String testFileKey, @Nullable String sourceFileUuid,
+    @Nullable Integer sourceFileLineNumber,
+    SearchOptions searchOptions) {
+    if (testUuid != null) {
+      return testIndex.searchByTestUuid(testUuid, searchOptions);
+    } else if (testFileUuid != null) {
+      return testIndex.searchByTestFileUuid(testFileUuid, searchOptions);
+    } else if (testFileKey != null) {
+      ComponentDto testFile;
+      DbSession dbSession = dbClient.openSession(false);
+      try {
+        testFile = dbClient.componentDao().getByKey(dbSession, testFileKey);
+      } finally {
+        MyBatis.closeQuietly(dbSession);
+      }
+      return testIndex.searchByTestFileUuid(testFile.uuid(), searchOptions);
+    } else {
```

Comment:
in this precise case, I shoudn't need to check that sourceFileUuid and sourceFileLineNumber are not null as I already have checked parameters conditions